common/src/primitives/merkle/tree.rs

```diff
@@ -0,0 +1,560 @@
+    }
+
+    /// Ensure the leaves indices are sorted and unique
+    fn is_sorted_and_unique(leaves_indices: &[u32]) -> bool {
+        let vals: BTreeSet<u32> = leaves_indices.iter().copied().collect();
+        let vals: Vec<u32> = vals.into_iter().collect();
+        vals == leaves_indices
+    }
+
+    /// Find adjacent leaves indices in a merkle tree
+    fn get_adjacent_indices_states(
+        leaves_indices: &[u32],
+    ) -> Result<Vec<AdjacentLeavesIndices>, MerkleTreeProofExtractionError> {
+        if !Self::is_sorted_and_unique(leaves_indices) {
+            return Err(
+                MerkleTreeProofExtractionError::UnsortedOrUniqueLeavesIndices(
+                    leaves_indices.to_vec(),
+                ),
+            );
+        }
+
+        let mut res = Vec::with_capacity(leaves_indices.len());
+
+        // we chain the windows with a max value to ensure we get the last element if it doesn't pair with the preceding value
+        let max_chain = std::iter::once(&u32::MAX);
+        for win in leaves_indices.iter().chain(max_chain).tuple_windows::<(&u32, &u32)>() {
+            let (a, b) = (*win.0, *win.1);
+
+            // In a tree, we expect elements to be adjacent if the first one has even index and the second one has index + 1.
+            if a % 2 == 0 && a + 1 == b {
+                res.push(AdjacentLeavesIndices::Together(a as usize, b as usize));
+            } else {
+                res.push(AdjacentLeavesIndices::Alone(a as usize));
+            }
+        }
+
+        Ok(res)
+    }
+
+    pub fn node_from_bottom(
+        &self,
+        level_from_bottom: usize,
+        index: usize,
+    ) -> Result<H256, MerkleTreeAccessError> {
+        let leaves_count = Self::leaves_count_from_tree_size(
+            NonZeroUsize::new(self.tree.len()).expect("By design, tree_size is always > 0"),
+        )
+        .expect("Failed to extract leaves count");
+
+        let level_count = leaves_count.trailing_zeros() as usize + 1;
+        if level_from_bottom >= level_count {
+            return Err(MerkleTreeAccessError::LevelOutOfRange(
+                self.tree.len(),
+                level_from_bottom,
+                index,
+            ));
+        }
+        let level_from_top = level_count - level_from_bottom;
+        let level_start = (self.tree.len() >> level_from_top) << level_from_top;
+        let index = level_start + index;
+        // TODO(PR): check index access
+        let index_size = usize::MAX;
+        if index >= index_size {
+            return Err(MerkleTreeAccessError::IndexOutOfRange(
+                self.tree.len(),
+                level_from_bottom,
+                index,
+            ));
+        }
+
+        Ok(self.tree[index])
+    }
+
+    /// Multi-proof of inclusion for a list of elements
+    pub fn multi_proof(
+        &self,
+        leaves_indices: &[u32],
+    ) -> Result<Vec<H256>, MerkleTreeProofExtractionError> {
```

Comment:
Returning some sort of `MerkleProof` type instead of plain `Vec<H256>` would be nice here.

Reply:
Sure... this is still a very rough thing... not even close to complete. I still don't even know how to represent the proof, and I assure you it's not just a vector.

Damn I'm seeing all these comments in my IDE and just can't help it...